### PR TITLE
Add elements support to the typography panel in global styles

### LINKED ISSUE #36546: Global Styles panel: Add elements support to “Typography” section
Opening this issue to discuss adding elements support to the “Typography” section of the Global Styles panel. This could potentially just be Text and Links at first.

### PR DESCRIPTION
**Pull request by youknowriad**
closes #36546 

This PR adds an element selector to the typography panel to match the proposed global styles design.
The panel is only enabled in the root of the global styles for the moment. (We don't have a clear way to say which blocks supports which element)

**Notes**

If we want to add headings there, I think we'd need a row per heading level if I'm not wrong. right @jorgefilipecosta @oandregal ?

### CODE REVIEW

**Comment by jasmussen** [comment]
Nice one. Before:
before

After looks excellent, except for the font size being inherited in the tiny little preview:

after

That'd be my only niggle. 

CC: @ciampo for the `Subtitle` subheading used 👌

**Comment by ciampo** [comment]
> CC: @ciampo for the `Subtitle` subheading used 👌

Thanks for the ping! The `Subtitle` component is a [level 2 `Heading`](https://github.com/WordPress/gutenberg/blob/d78b2714a8678bc3ea6893bda9863f331f32bfb9/packages/edit-site/src/components/global-styles/subtitle.js#L8) with some [additional custom CSS](https://github.com/WordPress/gutenberg/blob/892a93b3cfd02c3fb103818f24c3fd4158d7a742/packages/edit-site/src/components/global-styles/style.scss#L49-L54) applied. I've added this example to the list in #35464

File: packages/block-editor/src/hooks/typography.js [hunk 1]
```diff
@@ -231,7 +231,7 @@ export function TypographyPanel( props ) {
 	);
 }
 
-const hasTypographySupport = ( blockName ) => {
+export const hasTypographySupport = ( blockName ) => {
```
Comment (ntsekouras) [inline packages/block-editor/src/hooks/typography.js:234]: Why this change? You haven't used anywhere else this export.

File: packages/edit-site/src/components/global-styles/screen-typography-element.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import TypographyPanel from './typography-panel';
+import ScreenHeader from './header';
+
+const elements = {
+	text: {
+		description: __( 'Manage the fonts used on the site.' ),
+		title: __( 'Text' ),
+	},
+	link: {
+		description: __( 'Manage the fonts and typography used on the links.' ),
```
Comment (ntsekouras) [inline packages/edit-site/src/components/global-styles/screen-typography-element.js:18]: I guess these descriptions should be more consistent and change only the part `on the site|links`, no?

**Review by ntsekouras — COMMENTED** [review]
Thanks for the PR Riad!

I'd echo Joen about the preview text. After reading the issue I'd expect a preview card like this from  [this comment](https://github.com/WordPress/gutenberg/issues/36546#issuecomment-971671666):

132199312-a3610d4e-3b27-48b2-9e45-843090c9f830

Otherwise this looks good.

File: packages/block-editor/src/hooks/typography.js (hunk 1 above)
Comment (youknowriad) [reply packages/block-editor/src/hooks/typography.js:234]: Yeah, I changed my mind :)

File: packages/edit-site/src/components/global-styles/screen-typography-element.js (hunk 2 above)
Comment (youknowriad) [reply packages/edit-site/src/components/global-styles/screen-typography-element.js:18]: I'm happy to use any proposed descriptions here :)

**Comment by youknowriad** [comment]
@ntsekouras this PR didn't update the design of the typography panel itself, just the elements part.

**Comment by ntsekouras** [comment]
>@ntsekouras this PR didn't update the design of the typography panel itself, just the elements part.

To clarify the only part I was referring to, was this:
Screenshot 2021-11-22 at 7 51 04 PM

This is not a part of this PR?

**Comment by oandregal** [comment]
> If we want to add headings there, I think we'd need a row per heading level

Yeah, h1-h6 are different elements, so if we want to expose them all in the UI for the user to change we need to do it individually.

**Comment by youknowriad** [comment]
> This is not a part of this PR?

No, it's not but I can add it, it shouldn't be that hard.

File: packages/edit-site/src/components/global-styles/typography-panel.js [hunk 3]
```diff
@@ -54,8 +54,10 @@ function useHasLetterSpacingControl( name ) {
 	);
 }
 
-export default function TypographyPanel( { name } ) {
+export default function TypographyPanel( { name, element } ) {
```
Comment (oandregal) [inline packages/edit-site/src/components/global-styles/typography-panel.js:57]: Hey, I'd like us to revisit this conversation https://github.com/WordPress/gutenberg/pull/35264#discussion_r723221063

What are your thoughts about passing the underlying functions `useSetting`, `useStyle`, and `getSupportedGlobalStylesPanels` the parameters `path` and `context`? The context would contain any extra metadata we need (source, block name, element, etc). I've recently introduced this change for the server-side API to be future-proof in regards to addressing different templates down the road https://github.com/WordPress/gutenberg/pull/36610 

Not a blocker for this PR, although I'm now more convinced we should do it for 5.9 than I was a few months ago when we first discussed it.

Comment (youknowriad) [reply packages/edit-site/src/components/global-styles/typography-panel.js:57]: yep, I'm fine with that change but this is an internal API for the moment, so there's no real urgency yet.

**Comment by critterverse** [comment]
Nice! Super small note that the original designs indicate underlined text in the inline link preview. Is that something we could add here?

type-elements

**Comment by youknowriad** [comment]
> Nice! Super small note that the original designs indicate underlined text in the inline link preview. Is that something we could add here?

I feel like this should be a theme's decision, do we have a block support already to underline things? Maybe we should show it in the typography panel to allow users to enable/disable underline for texts.

**Comment by mtias** [comment]
> do we have a block support already to underline things?

There should be a text-decoration tool but it was not enabled as a cosmetic one for non-link items.

**Comment by carolinan** [comment]
Links are underlined by browser default, shouldn't they be underlined until the theme or user _removes_ them,
as opposed to being underlined only when the theme sets an underline style.

**Comment by youknowriad** [comment]
I've added the preview card and also forced links to be "underlined" for now until we add a text decoration control there.

**Comment by youknowriad** [comment]
What's missing here, can we get this iteration in?

**Comment by jasmussen** [comment]
To me this looks good to land as an iteration, except for one thing. We still have the font size inherited in this tiny little thumbnail:

Screenshot 2021-11-24 at 11 12 09

I think we need to just enforce a font size, say 13 or 14px there. We can still inherit the font and style.

**Comment by youknowriad** [comment]
Works for me, I'll update it.

**Comment by mtias** [comment]
Yes, the size is meant to be hard-coded in the small preview.

**Comment by youknowriad** [comment]
Updated and used 14px there.

**Comment by mtias** [comment]
Can we bring color as well if it exists for that element's preview?

**Comment by youknowriad** [comment]
@mtias sure 👍

**Comment by youknowriad** [comment]
Colors added (both background and text)

File: packages/edit-site/src/components/global-styles/screen-typography-element.js (hunk 2 above)
Comment (ntsekouras) [reply packages/edit-site/src/components/global-styles/screen-typography-element.js:18]: So I guess we should remove the export here?

**Review by ntsekouras — APPROVED** [review]
LGTM - thanks Riad!

**Comment by mtias** [comment]
Thanks! :)

File: packages/edit-site/src/components/global-styles/screen-typography-element.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+		title: __( 'Link' ),
```
Comment (mtias) [inline packages/edit-site/src/components/global-styles/screen-typography-element.js:19]: This should probably be plural

Comment (youknowriad) [reply packages/edit-site/src/components/global-styles/screen-typography-element.js:19]: https://github.com/WordPress/gutenberg/pull/36819

**Comment by briceduclos** [comment]
Will 5.9 also include a way to change the Headings as shown by @critterverse?
I think that changing the heading font properties is more common and useful than adjusting the link fonts.

> type-elements

**Comment by youknowriad** [comment]
@briceduclos not at the moment. FSE don't allow to change all headings at the same time, we can potentially have an entry per heading level, I'm waiting for more design input there.

**Comment by mtias** [comment]
We need to first establish what "headings" mean since it can span multiple blocks (heading, post title, archive title, comment title, etc). For that we need to first focus on the block API aspects of designating things as headings (which also includes table of contents, outline, etc). Once we have a plan there we can incorporate it as an element. Link and Text are already established and working.

**Comment by bradhogan** [comment]
Agree that changing headings from body copy is pretty important.  Any updates on this feature being included?

**Comment by poof86** [comment]
I see a + sign here to add more elements for custom styling, though i'm missing this in the editor atm. Will this be added? I think it would be a nice thing to have
